feat: Updated TabsList to no longer use ScrollView, but only rendering active tab

brianacnguyen opened the pull request:
## **Description**

This PR simplifies the `TabsList` component by removing complex ScrollView-based tab navigation in favor of a lightweight gesture-based approach.

**What is the reason for the change?**

The previous `TabsList` implementation had significant complexity that was unnecessary for the wallet page use case:
- Heavy ScrollView with horizontal paging and complex scroll handling
- Multiple refs, timeouts, and state management for scroll coordination
- Complex height measurement logic with debouncing and timing constants
- Adjacent tab preloading logic
- Over 1,600 lines of code across implementation and tests

This complexity caused:
- Difficult maintenance and debugging
- Performance overhead from scroll event handlers
- Conflicts with page-level scrolling in the wallet redesign
- Flaky tests due to timing-dependent behavior

**What is the improvement/solution?**

This PR dramatically simplifies TabsList by:

1. **Replacing ScrollView with gesture-based navigation** using `react-native-gesture-handler`
   - Pan gestures detect swipe left/right
   - Direct tab switching with `hidden` CSS class
   - No scroll event coordination needed

2. **Simplifying tab rendering**
   - Removed horizontal ScrollView and paging logic
   - Uses simple show/hide with `hidden` class
   - All tabs render in place, inactive tabs are hidden

3. **Streamlined caching**
   - Only loads the active tab on demand
   - Removed adjacent tab preloading
   - Removed complex timing coordination

4. **Reduced code complexity**
   - **389 lines removed** from TabsList.tsx
   - **1,357 lines removed** from tests
   - Removed refs for scroll coordination, timeout management, and measurement timers
   - Removed dependencies: `useTailwind`, `Dimensions`, scroll event types

5. **Improved gesture support**
   - Natural swipe thresholds (50px or 500 velocity)
   - Proper worklet execution with `react-native-reanimated`
   - Skips disabled tabs during swipe navigation

**Benefits:**
- ✅ Simpler, more maintainable code
- ✅ Better compatibility with page-level scrolling
- ✅ No scroll event overhead
- ✅ More reliable tests (no timing dependencies)
- ✅ Cleaner architecture for wallet page integration

## **Changelog**

CHANGELOG entry: Simplified TabsList component with gesture-based navigation

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/jira/software/c/projects/DSYS/boards/1888?search=tab&selectedIssue=DSYS-247

## **Manual testing steps**

```gherkin
Feature: Simplified TabsList navigation

  Scenario: user swipes between tabs
    Given TabsList with multiple tabs
    
    When user swipes left
    Then next tab displays
    And transition is smooth
    
    When user swipes right
    Then previous tab displays

  Scenario: user taps tabs directly
    Given TabsList with multiple tabs
    
    When user taps a tab button
    Then that tab becomes active immediately
    And only the active tab content loads

  Scenario: disabled tabs are skipped during swipe
    Given TabsList with some disabled tabs
    
    When user swipes through tabs
    Then disabled tabs are skipped
    And only enabled tabs are accessible

  Scenario: lazy loading of tab content
    Given TabsList with 4 tabs
    And user is on first tab
    
    When user switches to third tab
    Then third tab content loads
    And unused tabs remain unloaded
    And no preloading occurs
```

## **Screenshots/Recordings**

### **Before**

### **After**
https://github.com/user-attachments/assets/51537667-68f6-4ed3-8e15-a8a1b0262588

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable (✅ Tests updated to match simplified implementation)
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---


<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Refactors TabsList to use pan-gesture navigation and InteractionManager-driven lazy loading, rendering only loaded/active tabs and updating tests/snapshots accordingly.
> 
> - **Component `TabsList`**:
>   - Replace `ScrollView` paging and scroll handlers with `GestureDetector` pan gestures; remove scroll refs/timeouts and layout width logic.
>   - Add lazy loading via `InteractionManager.runAfterInteractions`; cache loaded tabs; render only active/loaded content; skip disabled tabs; preserve selection by `key` on children changes.
>   - Keep ref API (`goToTabIndex`, `getCurrentIndex`); pass through `tabsBarProps` and testIDs; simplify layout to a single `Box` container.
> - **Tests & Snapshots**:
>   - Rewrite tests to mock `InteractionManager`, verify deferred loading, cancellation, immediate loads for cached tabs, and gesture wrapper presence.
>   - Update/refocus tests on tab press behavior, disabled tabs handling, ref navigation, dynamic tabs by key, and edge cases (non-React children, initialActiveIndex).
>   - Refresh snapshots to reflect new structure without `ScrollView`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2d17260f597bffaf9c8f593d96b2eb70e51993bd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

georgewrmarshall commented on app/component-library/components-temp/Tabs/TabsList/TabsList.tsx:
```diff
@@ -116,34 +128,187 @@ const TabsList = forwardRef<TabsListRef, TabsListProps>(
       [enabledTabs],
     );
 
-    // Check if there are any enabled tabs and if current active tab is enabled
```
non-blocking: any reason to remove these comments?

georgewrmarshall reviewed (dismissed):
Great work getting the dynamic tab heights working! Approving to unblock delivery, but flagging critical technical debt for fast follow up.

The `TabsList` component is now managing 6+ timeout systems and multiple ref Maps, which raises potential memory management concerns:

* The cleanup effect runs expensive operations on every tab change
* Height measurement callbacks may hold onto stale closures after unmount
* Timers for scroll, load, and measurement aren’t coordinated and should be consolidated
* The height effect triggers 4+ times per tab switch due to overly broad dependencies
* A new Map is created on every height change, in a hot render path

The component has also grown in complexity. The height measurement logic alone spans 100+ lines, with tightly coupled effects that will be difficult to maintain or debug safely.

As part of the follow-up work with the @MetaMask/metamask-assets team, I'd recommend exploring ways to simplify the need for this level of complexity. It might also be worth bringing in the Margelo team for a review. They may have ideas on achieving this layout with a lighter implementation.

Prithpal-Sooriya reviewed (dismissed):
After a sync with @brianacnguyen, to unblock delivery, we are accepting the complex item size calculations.

We will a have a fast follow to clean this up - discussed in comments:
- https://github.com/MetaMask/metamask-mobile/pull/21822/files#r2473876069
- https://github.com/MetaMask/metamask-mobile/pull/21822/files#r2475981442

If possible can we get some QA testing on this before it gets merged, just to alieviate some concerns or any need to revert this?

Unik0rnMaggie commented:
Testing on the QA build from [here](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/2d71b163-902e-4eeb-9959-94af7fdbd8bd) , the View all NFTs button is missing on iOS.

Android build failed, so it cannot be tested


https://github.com/user-attachments/assets/274a124a-1e1f-48fd-ad4c-752cf79f1104

gambinish commented on app/components/UI/Perps/Views/PerpsTabView/PerpsTabView.tsx:
```diff
@@ -75,6 +80,17 @@ const PerpsTabView: React.FC<PerpsTabViewProps> = () => {
+      isHomepageRedesignV1Enabled && {
+        flexGrow: undefined,
+        minHeight: isInitialLoading ? 100 : undefined,
+      },
```
Why are we introducing the `isHomepageRedesignV1Enabled` feature flag here?

gambinish reviewed (dismissed):
PerpsTab seems to work fine with these changes. I do notice a behavior where the screen snaps to the view, when the scrolled to tab takes up less vertical space than the previous tab. Unsure if there's anything to do about this?

It's most noticeable when the empty states or "Retry Connection" components are rendered, and you are swiping to that tab from another long list where you have previously scrolled down.

https://github.com/user-attachments/assets/2d632f07-babd-4dd6-b4ef-376868e2c351

brianacnguyen committed:
fix: Defer TabsList content loading until after animation completes

- Use InteractionManager to defer content rendering until animations finish
- Prevents UI lag during tab transitions
- Already-loaded tabs still render immediately
- Properly cleanup interaction handles on component unmount

brianacnguyen committed:
test: Update TabsList tests for InteractionManager deferred loading

- Mock InteractionManager.runAfterInteractions for test determinism
- Update tests to handle deferred content loading behavior
- Replace timer-based tests with InteractionManager-based tests
- Add tests for interaction cancellation on rapid tab switching
- Add test for already-loaded tabs rendering immediately
- Update assertions for disabled tabs to check activeIndex instead of DOM
- Update snapshots to match new deferred loading behavior

brianacnguyen committed:
Merge branch 'main' into fix/tab-individual-height

Matt561 reviewed (dismissed):
Tests well on iOS simulator and Android physical device. Feel free to ping for re-approval once code coverage is fixed! 👍

brianacnguyen committed:
Updated tabslist test coverage

brianacnguyen committed:
Merge branch 'fix/tab-individual-height' of https://github.com/MetaMask/metamask-mobile into fix/tab-individual-height

brianacnguyen committed:
Merge branch 'main' of https://github.com/MetaMask/metamask-mobile into fix/tab-individual-height

brianacnguyen committed:
Merge branch 'main' into fix/tab-individual-height
